### ISSUE TITLE
Idiomatic way for `if` with external command

### ISSUE THREAD

**Issue by ineu** [nushell/nushell#11460]
### Related problem

There's a very common case in shell scripting:
```
if grep -q something file; then
   ...
fi
```
Alas, this use case is not properly documented in nu's `help if`: the help only shows examples of comparing numbers, which makes sense for a programming language but not for a shell.

The only way to do this in Nu I have found is:
```
if (grep -q something file | complete).exit_code == 0 {
   ...
}
```
which is very verbose and not intuitive.

So the questions are:
1. If this approach is indeed the only way? If so, it should be documented
2. If not, what is the idiomatic approach?
3. Can it be made simpler (a topic for discussion)? For example, there's the `^` operator which invokes external command. A similar operator could be added to change the subcommand's behavior to return boolean from the exit code, like this:
```
if @(grep -q something file) {
   ...
}
```
4. Does it make sense at all for a subcommand in `if` to return stdout? I.e. what is more common to check: the string output of a subcommand or the exit code?
```
if (grep something file) == "something" { ... }
if (grep -q something file | complete).exit_code == 0 { ... }
```
I tend to think the latter is more common and should be the default, but of course it's only an assumption

### Describe the solution you'd like

Make `if` + command more intuitive (or at least documented)

**Comment by fdncred** [comment]
You could write a helper function similar to this one.
```nushell
def grep-q [term file]: nothing -> bool {
  if (grep -q $term $file | complete).exit_code == 0 {
    true
  } else {
    false
  }
}
```
```nushell
❯ if (grep-q tool toolkit.nu) { echo 'found it' } else { echo 'not found' }
found it
```

**Comment by ineu** [comment]
Of course, but it's not about this specific case. I cannot write helper methods for all external commands :)
Another good example is the `test` command. Nu's `path` commands provide a subset of its capabilities, but not all, so sometimes you'd have to fall back to the `test` itself.
Yet another is `fd -q` to find a file in a filesystem, which exits with 1 if nothing was found.

**Comment by ineu** [comment]
Another option which just came to my mind is to add a flag to `run-external` which would return true/false based on the argument's exit code, i.e.
```
if (run-external --the-new-flag "grep" "-q" "term" "file") {
   ...
}
```
It could be the simplest thing to implement. Though annoying to use because of the quotes around each word.

**Comment by texastoland** [comment]
I struggled with this in nushell/nu_scripts#723 too. As you say it's quite pervasive in shell scripting.

**Comment by ineu** [comment]
This looks reasonable, however I'm not sure it will work:
```
~ ❌1 ❯ if ({ || true }) { echo foo}
Error: nu::shell::cant_convert

  × Can't convert to bool.
   ╭─[entry #4:1:1]
 1 │ if ({ || true }) { echo foo}
   ·    ──────┬──────
   ·          ╰── can't convert closure to bool
   ╰────
```

**Comment by fdncred** [comment]
> But in general, should the parser be changed 

I don't think so. I think it's working as designed. If you want to evaluate a sub-expression, it has to be enclosed in parens.
```nushell
❯ let closure = {|| true}
❯ if (do $closure) { "hi"} else { "hello"}
hi
```

**Comment by ineu** [comment]
Yes, that's why I'm saying it would require changes to the parser, so the first opening curly bracket won't be treated as "then" block. Not sure it's possible though because it would make syntax depend on semantics: since the removal of the mandatory `{ || }` syntax we can not tell if something in curly braces is a proc or a block ("then" in this case). Adding exceptions like "this is a proc if it follows `do` and a block otherwise" is not an option.

**Comment by cablehead** [comment]
A use case I have is:

```
curl -v -sN --unix-socket $"($store)/sock" $url 
    | lines | each { |x| $x | from json }
```

This fails quietly if the sock doesn't exist, or doesn't have an active listener. Perhaps external commands could emit an error on the pipeline after a none-zero exit code?

Bash would be a safer tool if `set -euo pipefile` were the defaults. I'm not sure how realistic it be to change this behavior for external commands now?

**Comment by cablehead** [comment]
Some good context https://www.sophiajt.com/exit-codes/ 🙏🏻 @fdncred

**Comment by cablehead** [comment]
Somewhat related: https://github.com/nushell/nushell/issues/5960

**Comment by NotTheDr01ds** [comment]
With #13515 the idiomatic way to handle this would be with a `try` statement:

```nushell
try {
  grep -q export spam.nu
  print "Found"
}
```

I've added https://github.com/nushell/nushell.github.io/issues/1848 to get this documented at some point.
